### PR TITLE
Tag partners in Drive comment notifications

### PR DESCRIPTION
**Pull request by pwnage101**
At a high level:
* Implements a new DriveApi function list_permissions_for_files.
* Refactors create_comments_for_files to allow different comments per file.
* Updates the retirement_partner_report script to add tags to comments.
* Adds "future" to requerements, one small backstep for mankind.

PLAT-2265

### CODE REVIEW

File: tubular/scripts/retirement_partner_report.py [hunk 1]
```diff
@@ -248,12 +248,31 @@ def _add_comments_to_files(config, file_ids):
...
+
+    partner_folders_to_permissions = drive.list_permissions_for_files(config['partner_folder_mapping'].values())
+
+    permissions = {
+        partner: partner_folders_to_permissions[config['partner_folder_mapping'][partner]]
+        for partner in file_ids
+    }
+
+    reader_emails = {
+        partner: [perm['emailAddress'] for perm in permissions[partner] if perm['role'] == 'reader']
```
Comment (bmedx) [inline tubular/scripts/retirement_partner_report.py]: Are we sure that the primary role for all partner contacts will always be `reader`?

File: tubular/google_api.py [hunk 2]
```diff
@@ -327,6 +327,83 @@ def create_comments_for_files(self, file_ids, content, fields='id'):
...
+        file_id_to_request_id = dict(zip(
+            file_ids,
+            (str(n) for n in count()),
+        ))
+        # Create a flipped mapping for convenience.
+        request_id_to_file_id = {v: k for k, v in iteritems(file_id_to_request_id)}
+
+        def callback(request_id, response, exception):  # pylint: disable=unused-argument,missing-docstring
+            file_id = request_id_to_file_id[request_id]
+            responses[file_id] = response['permissions']
```
Comment (bmedx) [inline tubular/google_api.py]: Will the `permissions` key be present if there is an exception? Similarly  for the callback in `_create_comments_for_files` - though the responses seem to eventually be thrown out there, we're just logging exception responses and ignoring them?

File: tubular/tests/test_google_api.py [hunk 3]
```diff
@@ -641,4 +646,45 @@ def test_comment_files_with_duplicate_file(self, mock_from_service_account_file)
         ])
         test_client = DriveApi('non-existent-secrets.json', http=http_mock_sequence)
         with self.assertRaises(ValueError):
-            test_client.create_comments_for_files(fake_file_ids, 'some comment message')
+            test_client.create_comments_for_files(list(zip(fake_file_ids, cycle(['some comment message']))))
+
+    @patch('tubular.google_api.service_account.Credentials.from_service_account_file', return_value=None)
+    def test_list_permissions_success(self, mock_from_service_account_file):  # pylint: disable=unused-argument
+        """
+        Test normal case for listing permissions on files.
```
Comment (bmedx) [inline tubular/tests/test_google_api.py:654]: Is there a case for exercising the error conditions to make sure they're being eaten and/or returned / raised correctly?

File: tubular/google_api.py (hunk 2 above)
Comment (pwnage101) [reply tubular/google_api.py]: I forgot to do any reasonable error handling, reworked this section to hard fail on any error.

File: tubular/tests/test_google_api.py (hunk 3 above)
Comment (pwnage101) [reply tubular/tests/test_google_api.py:654]: added a new test to cover the an error case (file not found).

File: tubular/scripts/retirement_partner_report.py (hunk 1 above)
Comment (pwnage101) [reply tubular/scripts/retirement_partner_report.py]: I'll hopefully have an answer for this later.

Comment (pwnage101) [reply tubular/scripts/retirement_partner_report.py]: the answer: filter out emails ending with '@edx.org'

**Comment by pwnage101** [comment]
Ready for re-review. Just added email filtering based on whether or not the domain is "edx.org".

I know we discussed we may not even be doing these notifications, but may as well merge it since it's all here now.

File: tubular/google_api.py [hunk 4]
```diff
@@ -327,6 +334,85 @@ def create_comments_for_files(self, file_ids, content, fields='id'):
                 For some non-retryable 4xx or 5xx error.  See the full list here:
                 https://developers.google.com/drive/api/v3/handle-errors
         """
+        file_ids, _ = zip(*file_ids_and_content)
+        if len(set(file_ids)) != len(file_ids):
+            raise ValueError('Duplicates detected in the file_ids_and_content list.')
```
Comment (doctoryes) [inline tubular/google_api.py:339]: Nit: Would it be worth including the duplicates here in the exception message? Looks like you could use:
https://docs.python.org/2/library/stdtypes.html#frozenset.symmetric_difference

File: tubular/scripts/retirement_partner_report.py [hunk 5]
```diff
@@ -207,12 +213,40 @@ def _add_comments_to_files(config, file_ids):
...
+        partner: partner_folders_to_permissions[config['partner_folder_mapping'][partner]]
+        for partner in file_ids
+    }
+
+    # throw out all edx.org addresses, and flatten the permissions dicts to just the email:
+    external_emails = {
+        partner: [
+            perm['emailAddress']
+            for perm in permissions[partner]
+            if not perm['emailAddress'].endswith('@edx.org')
```
Comment (doctoryes) [inline tubular/scripts/retirement_partner_report.py]: Nit: I don't *think* it should be an issue here, but perhaps:
`if not perm['emailAddress'].lower().endswith('@edx.org')` ?

**Review by bmedx — APPROVED** [review]

**Review by doctoryes — APPROVED** [review]

File: tubular/scripts/retirement_partner_report.py (hunk 5 above)
Comment (pwnage101) [reply tubular/scripts/retirement_partner_report.py]: hah

File: tubular/google_api.py (hunk 4 above)
Comment (pwnage101) [reply tubular/google_api.py:339]: wouldn't work due to operator only working with sets.  I'm sure there's something else out there.

**Pull request merged**